[IRGen] Improvements to function type metadata

xedin opened the pull request:
Currently only single 'inout' flag has been encoded into function
metadata, these changes extend function metadata to support up to
32 flags per parameter.

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[IRGen] Add parameter flags to function type metadata

Currently only single 'inout' flag has been encoded into function
metadata, these changes extend function metadata to support up to
32 flags per parameter.

xedin commented:
@rjmccall Do changes so far make sense to you? It seems like the next step would be to modify `MetadataCache` to support this new flags to unique function keys. And at the end do couple of optimizations - don't encode flags for parameters which don't have any etc.

DougGregor commented on include/swift/AST/Types.h:
```diff
@@ -1450,11 +1450,26 @@ class ParameterTypeFlags {
+    return ParameterTypeFlags()
+        .withVariadic(value & Variadic)
+        .withAutoClosure(value & AutoClosure)
+        .withEscaping(value & Escaping)
```
I don't feel like autoclosure or escaping belong in the parameter flags, but that concern is more-or-less independent of your improvements to function type metadata.

xedin replied:
@DougGregor I think I can mask them out from the metadata with ease, do you think the approach here is viable overall?

rjmccall commented on include/swift/Runtime/Metadata.h:
```diff
@@ -1760,7 +1758,14 @@ struct TargetFunctionTypeMetadata : public TargetMetadata<Runtime> {
+  uint32_t getParameterFlags(unsigned index) const {
+    auto numParams = getNumArguments();
+    assert(index < numParams);
+    auto *flags = reinterpret_cast<const uint32_t *>(this + 1 + numParams);
```
This pointer computation seems very wrong; the ``+ numParams`` is scaled by sizeof(FunctionTypeMetadata).

Also, I'm not sure I understand why you're being incremental about changing the ABI here, at least as regards the layout of FunctionTypeMetadata.  Is there a good reason not to design and implement the desired ABI at this point, since you're revising the ABI anyway?  Why force someone else to come by and do all these same updates again?

xedin replied:
I will redo the pointer but I don’t understand what do you mean by incremental changes here, can you please elaborate? I am thinking of doing everything related to function metadata ABI, this is just a first changes wrt the layout I wanted input on...

rjmccall replied:
This seems to be implementing a flat, always-present array.  The things we talked about:
  - maybe not representing this as an array at all
  - maybe only including elements in the array if they have flags
  - or at least allowing the array to be optional

Every change you make here has to be quite involved in terms of e.g. updating all the runtime functions and the compiler's calls to them and how MetadataReader interprets the runtime structures, so I'm questioning whether it really makes sense to do all these changes in separate patches.

xedin replied:
Ah sure, I understand. I'm planning to do all this next as one step and update existing commit, just wanted to make sure that I haven't done something obviously terrible since I'm pretty new to LLVM...

rjmccall commented on include/swift/Runtime/Metadata.h:
```diff
@@ -2642,24 +2647,20 @@ swift_getFunctionTypeMetadata(const void *flagsArgsAndResult[]);
-swift_getFunctionTypeMetadata1(FunctionTypeFlags flags,
-                               const void *arg0,
+swift_getFunctionTypeMetadata1(FunctionTypeFlags flags, const void *arg0,
+                               const void *paramFlags,
```
If we're no longer mangling things into the low bits of the parameter types, we can just use 'const Metadata *' instead of 'void *', since we're not trying to encourage caution.

Why is paramFlags a void*?

getFunctionTypeMetadataN is a convenience function for common cases.  Do we have any reason to think that parameter type flags are a common enough case to justify making these functions less convenient?

You also need to update all the getFooFunctionTypeMetadata functions below this, or at least their most-general cases.

xedin replied:
I wasn’t sure what to you for parameter flags here, should it be `uint32_t *` instead? I think parameter flags for single parameter are pretty common, I’ve seen examples in compatibility suite.

rjmccall replied:
I don't doubt that they occur, just that they occur in enough frequency to warrant adding extra requirements to the common case.  Can you get counts here?

xedin replied:
I've found 87 occurrences of functions with `inout` and 1-3 arguments in source compatibility projects, most of them in `GRDB` project. I'm pretty sure there is going to be much more with `__shared` when that's available... I'm also curious - if not array of parameters how else should we track flags in this case, with individual arguments?

rjmccall replied:
Do you happen to have the full histogram for the source compatibility suite?  Specifically I'd like to know the cross table of (convention) x (argument count) x (has any flags).

I agree that __shared/__owned will make flags more common.  (We'll have to decide which is the default, and obviously it should track the language decision because the language decision will make the default much more common than the other.)

Our original decision to go up to 3 was not based on any particular insight that I remember.  The purpose of these functions is to optimize code size for common cases, and setting up a call with 7 arguments is already a ton of code, so the benefit might be negligible.  Maybe we'll want to provide specializations for (1,false), (2, false), (3, false) but just (1, true).  Certainly it seems questionable whether we should do these common-path entrypoints for each of the secondary ``@convention``s; maybe we should just have common-path entrypoints that take the convention as an argument.  Anyway, data is the key here.

xedin replied:
Unfortunately no, I just made a grep over all of the projects in the `project_cache` folder, but I will try to get all of the info with some effort.

> Maybe we'll want to provide specializations for (1,false), (2, false), (3, false) but just (1, true)...

You mean like `get_swiftFunctionTypeMetadata1InOut`, `get_swiftFunctionTypeMetadata1Shared` etc. ?

I'm actually wondering how many more flags are there possibly we'd want to add? It seems like from `MetadataCache` perspective it would be much easier to just make a custom metadata type ref for parameters with flags, which would encode flags as part of type metadata body but I'm not sure if that's a good idea and how feasible that would be? It seems at least it would be much easier to unique - no flags - pointer to metadata of type, at least one flag - pointer to metadata ref for "parameter" with includes metadata ref for type + uint32_t for flags, WDYT?

rjmccall replied:
Oh, are you just grepping for typealiases of function type or something?  We don't actually need to build type metadata for every function type we see.  Although I guess it's a significantly larger population, and I have no reason to think it's likely to be unrepresentative of the population of function types that we need metadata for.

RE: (1, false), I meant something more like swift_getFunctionTypeMetadataWithFlags1 vs. something that just assumed the default flags.

RE: MetadataCache: if MetadataCache makes it awkward to handle heterogenous key data, we can easily just drop down to SimpleGlobalCache (i.e. ConcurrentMap) the same way that several of the other caches do.  MetadataCache is primarily there to provide additional locking so that we can ensure that the more expensive metadata kinds are uniquely initialized, and function-type metadata don't need that.

xedin replied:
> Oh, are you just grepping for typealiases of function type or something?

Yes, that's exactly what I was doing, I wish we had some tools to generate reports like that but I'm afraid we don't, or maybe I just don't know about them. My thinking was that if there aren't many functions with `inout` (overall) in the project then and even fewer are going to require metadata generation that says something about distribution...

RE: `swift_getFunctionTypeMetadataWithFlags1` - Sure! I got that idea as well, just wasn't sure if you are proposing something even more sophisticated... I think it makes sense because many 1/2/3 parameter functions are likely not to have any flags, judging by compatibility suite. Which brings us to question about having separate "function parameter metadata" which is going to include pointer to type + uint32_t for flags and not going to require any changes to `swift_getFunctionTypeMetadata*` functions and `MetadataCache`, WDYT?

RE: MetadataCache - Ah, I see, that make sense to me!

rjmccall replied:
I think having a separate metadata object for each parameter would be rather gratuitously wasteful, and I can't think of anything that would benefit from it.

xedin replied:
Not always but only if there are flags present, if that’s not a very common occurance it’s not going to waste too much space, right?

rjmccall replied:
I think there's some value in having the layout of the component types not vary that drastically based on whether flags are present.  This is why I was suggesting having separate arrays.

xedin replied:
I totally agree! The reason why I bring up this parameter type ref metadata idea is because it makes it much easier to work with function metadata and doesn't require any specific optimizations, like in case of the array of flags, we'd like to keep only flags for parameters which have them, and don't use any additional place, such most likely would be more complicated comparing to extracting a single `uint32_t` from metadata ref and this approach wouldn't require adding even more optimized calls for 1/2/3 parameter functions with and w/e flags...

xedin committed:
[IRGen] Switch function metdata to use `Metdata *` for parameter references

xedin commented:
@rjmccall Can you please take another look? I've updated everything according to our discussion - added new `WithFlags` methods (and made them used if there are no parameters with flags), fixed parameters to be `const Metadata *`, plus a bug related to casting. I think the only remaining thing is to fix function metadata cache to respect new flags...

rjmccall commented on include/swift/AST/Types.h:
```diff
@@ -1450,11 +1450,26 @@ class ParameterTypeFlags {
+  ParameterTypeFlags withAutoClosure(bool isAutoClosure) const {
+    return ParameterTypeFlags(isAutoClosure
+                                  ? value | ParameterTypeFlags::AutoClosure
+                                  : value - ParameterTypeFlags::AutoClosure);
```
Is this an actual int subtraction?  You can't do that safely unless you're asserting that the flag is already set in 'value'.

rjmccall replied:
This comment is still applicable.

xedin replied:
I overlooked that, it's not longer necessary for function metadata anyway, so I'll just remove all of the methods I've added to `ParameterTypeFlags`.

xedin replied:
And I've looked it up - it's actually not a int substitution, but rather an `OptionSet` operation, so it should be safe to do that.

rjmccall commented on include/swift/Remote/MetadataReader.h:
```diff
@@ -812,22 +813,17 @@ class MetadataReader {
-        FlaggedArgumentAddress &= ~InOutMask;
-        if (auto ParamTypeRef = readTypeFromMetadata(FlaggedArgumentAddress)) {
           Param.setType(ParamTypeRef);
+          Param.setFlags(ParameterTypeFlags::fromRaw(ParameterFlags));
```
It is really not okay to assume that the AST's ParameterTypeFlags type matches the runtime-metadata type.  Please make these separate types with their own internal representation of all the various flags.  I'm sure I asked you to do this before.

xedin replied:
Ok, will do.

rjmccall commented on include/swift/Runtime/Metadata.h:
```diff
@@ -2643,23 +2649,47 @@ swift_getFunctionTypeMetadata(const void *flagsArgsAndResult[]);
 const FunctionTypeMetadata *
+swift_getFunctionTypeMetadata1WithFlags(FunctionTypeFlags flags,
+                                        const Metadata *arg0,
+                                        const uint32_t *paramFlags,
```
This should just take a uint32_t, not a pointer.

xedin replied:
Ok, will do.

rjmccall commented on include/swift/Runtime/Metadata.h:
```diff
@@ -2643,23 +2649,47 @@ swift_getFunctionTypeMetadata(const void *flagsArgsAndResult[]);
+swift_getFunctionTypeMetadata2WithFlags(FunctionTypeFlags flags,
+                                        const Metadata *arg0,
+                                        const Metadata *arg1,
+                                        const uint32_t *paramFlags,
```
This should take two uint32_t's.

xedin replied:
If we are going to go this route I going to re-arrange it so it’s param1, flags1, param2, flags2 etc.

rjmccall replied:
Yes, that seems reasonable.

rjmccall commented on include/swift/Runtime/Metadata.h:
```diff
@@ -2643,23 +2649,47 @@ swift_getFunctionTypeMetadata(const void *flagsArgsAndResult[]);
+                                                const Metadata *arg1,
+                                                const Metadata *arg2,
+                                                const uint32_t *paramFlags,
+                                                const Metadata *resultMetadata);
```
Are you planning to collect that histogram of actual uses and then figure out what flags we really want in a follow-up?

Regardless, you should follow the same pattern here and take three uint32_t's.  Only the most general entrypoint should take it indirectly.

rjmccall commented on stdlib/public/runtime/Metadata.cpp:
```diff
@@ -321,14 +322,22 @@ class FunctionCacheEntry {
+      auto numArguments = getFlags().getNumArguments();
+      assert(index < numArguments);
+      auto *flags = reinterpret_cast<const uint32_t *>(
+          FlagsArgsAndResult[numArguments + 1]);
```
Wait, are you storing a pointer in the metadata?  What is this pointer pointing to?  Is the assumption that callers of getFunctionTypeMetadata will permanently allocate a flags array?

xedin replied:
Currently I have it created as constant global array and passed as a pointer, I think it makes sense to copy or re-arrange it somehow when stored in the cache, I just didn’t do that part yet...

xedin replied:
If I wanted to avoid array allocation and just encode flags inline after each parameter, how should I go avoid storing uint32_t as void*?

rjmccall replied:
I would suggest storing it as two separate arrays in the metadata object, one after the other.

xedin replied:
Sorry but I don’t think that I fully understand what you mean, I am talking about void ** we pass to `swift_getFunctionTypeMetadata`, do you mean the same thing by metadata object? The reason I decided to go with this const array is because it’s easy to pass it in, but I would really rather do it as param1,flags1, etc. here the same way as 1/2/3 functions are going to be. I just don’t know how to encode uint32_t value as an element of void ** ...

rjmccall replied:
Well, you could extend it to intptr_t and then cast, which is easy enough to do in LLVM.  But I wasn't thinking about the fact that swift_getFunctionTypeMetadata has such an opaque interface.  Frankly, I'm not sure why it does, vs. taking the FunctionTypeFlags and a types (and now a flags) pointer.

xedin replied:
So here is how I am thinking of encoding arbitrary parameters since it’s easy to use uint32_t: represent each of the parameter’s metadata as void * with last bit set to 1 if it has flags, 0 otherwise, format is going to be paramN, flagsN?, ... if the last bit is not set consider next pointer in the array to be new parameter, otherwise it’s uint3_t representing flags for current parameter, WDYT?

rjmccall replied:
Do you mean in the interface to swift_getFunctionTypeMetadata or do you mean in the storage in a FunctionTypeMetadata?  I tend to think that separate arrays, where the flags array is optional, is a better representation overall for both cases.

rjmccall replied:
In the metadata, we can store whether there are any parameter flags in the FunctionTypeFlags.

xedin replied:
I mean as swift_getFunctionTypeMetadata argument, but this might be beneficial for cache too because it aways storing a bunch of zeros in the array.

xedin commented on include/swift/Runtime/Metadata.h:
```diff
@@ -1760,7 +1758,15 @@ struct TargetFunctionTypeMetadata : public TargetMetadata<Runtime> {
   TargetPointer<Runtime, const Argument> getArguments() const {
     return reinterpret_cast<TargetPointer<Runtime, const Argument>>(this + 1);
```
@rjmccall I'm working on the changes to separate parameters from flags when stored in the entry of the cache but I need some help understanding how to do that, especially what `this + 1` really means here? I've read comment in `SimpleGlobalCache` which talks about how `getExtraAllocationSize` is used, so here is what I think I need to do to make it work, please correct me if I'm wrong:

1. Modify `getExtraAllocationSize` so it tail-allocates enough space for array of `uint32_t` equal to number of parameters like this

```
static size_t getExtraAllocationSize(Key key) {
    auto numParams = key.getFlags().getNumArguments();
    return key.getFlags().getNumArguments() * (sizeof(FunctionTypeMetadata::Argument) + sizeof(uint32_t));
  }
```
2. Add a method to `FunctionCacheEntry`:

```
  TargetPointer<Runtime, uint32_t *> getParameterFlags() {
    return reinterpret_cast<TargetPointer<Runtime, uint32_t *>>(__address here__);
  }
```

Should __address here__ be `this + (numParams * sizeof(TargetFunctionTypeMetadata::Argument)) + 1` ?

rjmccall replied:
Remember that pointer arithmetic in C is scaled by the size of the pointee type.  ``this + 1`` is an idiom used when you allocate extra memory after an object ("extra" meaning "not just ``sizeof(T)``"), because it very conveniently creates a pointer to that extra memory without needing to write out the type of ``this``.  You then generally need to cast that pointer to some other type in order to do further pointer arithmetic.

So address probably needs to be ``reinterpret_cast<char*>(this + 1) + (numParams * sizeof(TargetFunctionTypeMetadata::Argument))``.

rjmccall replied:
Or, alternatively, ``reinterpret_cast<Argument*>(this + 1) + numParams``.

xedin replied:
Thank you!

xedin committed:
[IRGen] Fix function metadata endpoints to use separate parameter/flags arguments

Switch most general endpoint to be `flags, parameters, parameterFlags, result`,
instead of opaque `void **`, more specialized ones to use follow argument scheme:
`flags, param0, [flags0], ..., paramN, [flagsN], result` and store parameter/flags
information separately in `FunctionCacheEntry::{Key, Data}` as well.

rjmccall commented on include/swift/Remote/MetadataReader.h:
```diff
@@ -812,22 +813,17 @@ class MetadataReader {
       for (StoredPointer i = 0; i < Function->getNumArguments(); ++i,
            ArgumentAddress += sizeof(StoredPointer)) {
-        StoredPointer FlaggedArgumentAddress;
+        StoredPointer ParameterAddress;
```
I would call this ParamMetadata or something like that.

xedin committed:
[ABI/IRGen] Add custom function parameter flags representation for metadata use

xedin commented:
@rjmccall Does this look better? It's not yet complete but 98% done I think, I made it so parameters and flags are stored separately and flags are encoded independently from AST and converted when needed.

rjmccall commented on include/swift/Runtime/Metadata.h:
```diff
@@ -1760,20 +1758,43 @@ struct TargetFunctionTypeMetadata : public TargetMetadata<Runtime> {
-  
+
+  void setParameterFlags(unsigned index, ParameterFlags flags) {
+    assert(index <= getNumArguments());
```
Just ``<``, not ``<=``.

xedin replied:
Ah wow :(

xedin replied:
I should stop doing coding at night :)

rjmccall commented on include/swift/Runtime/Metadata.h:
```diff
@@ -1760,20 +1758,43 @@ struct TargetFunctionTypeMetadata : public TargetMetadata<Runtime> {
+  }
+
+  ParameterFlags getParameterFlags(unsigned index) const {
+    assert(index <= getNumArguments());
```
Same thing here.

rjmccall commented on include/swift/Remote/MetadataReader.h:
```diff
@@ -806,28 +806,35 @@ class MetadataReader {
+
+      for (StoredPointer i = 0; i < numParameters; ++i,
+                         ArgumentAddress += sizeof(StoredPointer),
+                         ParameterFlagsAddress += sizeof(StoredPointer)) {
```
I believe you need to be adding sizeof(uint32_t) here.

But in general, none of this should be necessary, since it's not reading memory outside of the true bounds of the FunctionTypeMetadata.  You just need to update the Function case in readMetadata so that it computes the true bounds correctly, just like how it's done for Tuples.  If you do that, you should be able to just use the normal access functions on TargetFunctionTypeMetadata (as long as they don't assert anything you haven't already checked!).

xedin replied:
I think I understand now what you are talking about - it's done for parameters this way because we need to read a point to metadata itself from the argument position, but for flags we don't have to do that because they are values themselves. Thanks again!

rjmccall commented on include/swift/Runtime/Metadata.h:
```diff
@@ -2638,28 +2659,58 @@ swift_getGenericWitnessTable(GenericWitnessTable *genericTable,
-                               const Metadata *resultMetadata);
+swift_getFunctionTypeMetadata2WithFlags(FunctionTypeFlags flags,
+                                        const Metadata *arg0,
+                                        const uint32_t flags0,
```
Don't const-qualify a parameter like this.  Also, this should be ParameterTypeFlags, right?

xedin replied:
Got it!

xedin replied:
I think it's fine to make it `ParameterFlags` for 1/2/3 functions but should it be `const ParameterFlags *` for the general case too?

rjmccall replied:
The way I see is it that the use of a target-specific type is specifically just for passing these values as direct arguments to the runtime functions.  Arguably that implies that we should make the enums (``{Function,Parameter}TypeFlags``) non-templated and just make the runtime functions take a size_t with a comment about what enum they really are.  The place where we take an array of them by reference should definitely use a pointer to a 32-bit type.

xedin replied:
Alright, this means that my changes are correct :)

rjmccall commented on include/swift/ABI/MetadataValues.h:
```diff
@@ -606,6 +615,55 @@ class TargetFunctionTypeFlags {
 };
 using FunctionTypeFlags = TargetFunctionTypeFlags<size_t>;
 
+template <typename int_type> class TargetParameterTypeFlags {
```
We seem to be assuming that the backing storage for this is always a uint32_t.  I think it's still maybe a good idea to have it be templated, because it solves a lot of annoying calling-convention portability concerns to always pass a parameter as a full word instead of a smaller integer, but I think it merits a comment about why we're doing it.

rjmccall reviewed (commented):
Yes, I'm much happier with the basic approach here, thank you.  We can revisit the question of which specialized getFunctionTypeMetadata functions to provide in a follow-up patch.  I've left some comments, and I'm sure I'll have more with the complete patch, but please proceed with this design.

xedin commented on include/swift/ABI/MetadataValues.h:
```diff
@@ -551,9 +551,10 @@ template <typename int_type>
+    ConventionMask = 0x0F000000U,
+    ConventionShift = 24U,
+    ThrowsMask = 0x10000000U,
+    ParamFlagsMask = 0x01000000U,
```
Sorry, this is what `git clang-format` does, I've tried to fix indentation of enums where possible but failed in some places still :(

xedin replied:
Will do the changes to size of the flags after I'm done with the flags and other changes, if that's ok?

xedin commented on include/swift/ABI/MetadataValues.h:
```diff
@@ -606,6 +615,55 @@ class TargetFunctionTypeFlags {
+    InOutMask    = 0x10000000U,
+    SharedMask   = 0x01000000U,
+    VariadicMask = 0x00100000U,
+  };
```
I've just looked at some other places, and I didn't know that it's more expensive, will fix, thanks!

rjmccall replied:
It's pretty minor, but yes, it is often the case that you can do something with a very small integer in one instruction that takes multiple instructions with a larger integer, so if it's otherwise an arbitrary choice, prefer to use smaller integers for more important flags.

xedin replied:
Understood, thanks!

xedin commented on include/swift/Runtime/Metadata.h:
```diff
@@ -1745,11 +1745,10 @@ using EnumMetadata = TargetEnumMetadata<InProcess>;
+  using Argument = const TargetMetadata<Runtime> *;
 
   TargetFunctionTypeFlags<StoredSize> Flags;
+  TargetPointer<Runtime, const uint32_t> ParamFlags;
```
@rjmccall Just wanted to double check if that's what you were talking about when referring to how tuple metadata is done?

xedin committed:
[ABI/Runtime] NFC: Rename function metadata arguments to parameters

xedin commented:
Looks like after I've added new `WithFlags` methods I've broke something related to Objective-C but I have no clue how to debug that, if I define function with parameter flags now and try to run the code process is crashing with:

```
0  swift                    0x000000010ab5af7c llvm::sys::PrintStackTrace(llvm::raw_ostream&) + 60
1  swift                    0x000000010ab5b579 PrintStackTraceSignalHandler(void*) + 25
2  swift                    0x000000010ab56d39 llvm::sys::RunSignalHandlers() + 425
3  swift                    0x000000010ab5b9d2 SignalHandler(int) + 354
4  libsystem_platform.dylib 0x00007fff613e4f5a _sigtramp + 26
5  libsystem_platform.dylib 0x000000011d631b5a _sigtramp + 3156528154
6  libswiftCore.dylib       0x0000000117319c45 swift::metadataimpl::ObjCRetainableBox::retain(void*) + 21
7  libswiftCore.dylib       0x0000000117319c1c swift::metadataimpl::RetainableBoxBase<swift::metadataimpl::ObjCRetainableBox, void*>::initializeWithCopy(void**, void**) + 28
8  libswiftCore.dylib       0x00000001173179c4 swift::metadataimpl::ValueWitnesses<swift::metadataimpl::ObjCRetainableBox>::initializeWithCopy(swift::OpaqueValue*, swift::OpaqueValue*, swift::TargetMetadata<swift::InProcess> const*) + 36
9  libswiftCore.dylib       0x0000000117317960 swift::metadataimpl::BufferValueWitnesses<swift::metadataimpl::ValueWitnesses<swift::metadataimpl::ObjCRetainableBox>, 8ul, 8ul, (swift::metadataimpl::FixedPacking)1>::initializeBufferWithCopyOfBuffer(swift::ValueBuffer*, swift::ValueBuffer*, swift::TargetMetadata<swift::InProcess> const*) + 48
10 libswiftCore.dylib       0x0000000116b84194 swift::metadataimpl::BufferValueWitnesses<swift::metadataimpl::ValueWitnesses<swift::metadataimpl::ObjCRetainableBox>, 8ul, 8ul, (swift::metadataimpl::FixedPacking)1>::initializeBufferWithCopyOfBuffer(swift::ValueBuffer*, swift::ValueBuffer*, swift::TargetMetadata<swift::InProcess> const*) + 4287023204
11 swift                    0x00000001047578de llvm::MCJIT::runFunction(llvm::Function*, llvm::ArrayRef<llvm::GenericValue>) + 1358
12 swift                    0x0000000104763da2 llvm::ExecutionEngine::runFunctionAsMain(llvm::Function*, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&, char const* const*) + 3650
13 swift                    0x00000001030307b7 swift::RunImmediately(swift::CompilerInstance&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&, swift::IRGenOptions&, swift::SILOptions const&) + 7271
14 swift                    0x0000000102f95fea performCompile(swift::CompilerInstance&, swift::CompilerInvocation&, llvm::ArrayRef<char const*>, int&, swift::FrontendObserver*, swift::UnifiedStatsReporter*) + 25866
15 swift                    0x0000000102f8d5b2 swift::performFrontend(llvm::ArrayRef<char const*>, char const*, void*, swift::FrontendObserver*) + 8882
16 swift                    0x0000000102eb9573 main + 4819
17 libdyld.dylib            0x00007fff61164145 start + 1
```

@rjmccall Any ideas how should I approach this?

rjmccall commented on include/swift/Remote/MetadataReader.h:
```diff
@@ -838,8 +832,10 @@ class MetadataReader {
+                       .withThrows(Function->throws())
+                       .withParameterFlags(Function->hasParameterFlags());
       auto BuiltFunction =
           Builder.createFunctionType(Parameters, Result, flags);
```
I think you'll need to change the createFunctionType API to also take parameter flags, which probably means remapping them to the AST flags.

Also, what I mean about TupleMetadata is that you should modify the ``MetadataKind::Function`` case in ``readMetadata`` to start by reading the FunctionTypeMetadata and then use that to figure out the complete size to read, like the case for ``MetadataKind::Tuple`` does.

xedin replied:
Ah ok, I think I understand know, I thought you meant treat flags like labels are treated in case of tuple...

Regarding `createFunctionType` API - each of the elements in `Parameters` already has flags attached to it, so I think flags are covered that way?...

rjmccall replied:
Oh, yes, I see, that makes sense.

rjmccall commented:
I would guess that your bug is that you're scribbling out of bounds of the metadata object somehow.

xedin committed:
[RemoteAST] Fix metadata reader to properly read function metadata

xedin commented:
@rjmccall I think I've made all of the changes required to make it work, please take a look one more time, thanks!

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please test source compatibility

xedin committed:
[Casting] NFC: add some more tests for functions with parameter flags

xedin commented:
@swift-ci please clean test

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please test OS X platform

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test

rjmccall reviewed (commented):
Okay, in addition to the change I request above, it looks like you're still on the hook for two things:

- Changing the layout and size of the different flags as discussed.
- Coming up with good data for which convenience accessors we should actually provide.

But it's fine to handle those in follow-ups.

xedin commented:
@rjmccall Yes, I was thinking to do that after this changes are merged as a follow up PR at least for function flags.

xedin committed:
[AST] NFC: Remote unused accessor methods from `ParameterTypeFlags`

xedin commented on lib/IRGen/GenMeta.cpp:
```diff
@@ -759,78 +748,136 @@ namespace {
-            Address argPtr = IGF.Builder.CreateStructGEP(buffer, i + 1,
+
+          ConstantInitBuilder paramFlags(IGF.IGM);
+          auto flagsArr = paramFlags.beginArray();
```
@rjmccall One last question I have is if we should keep flags as constant array of stack allocated one? After digging into all this more it seems like we might better off with stack allocated array flags the same way as for parameters because we don't keep them around that long anyway, WDYT?

rjmccall replied:
It is almost certainly less overall binary size to use a const global array than to create one fresh on the stack.

xedin replied:
Ok, sounds good, I'll leave it as is then! Thanks again!

xedin commented:
@swift-ci please source compatibility

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please source compatibility

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please test source compatibility

rjmccall commented:
Congrats!

xedin commented:
Thanks, John! I will make couple of follow-up PRs as we discussed.

xedin commented:
@rjmccall looks like it was too early to celebrate since this changes broke LLDB :( Do you have any idea where should I start in that project to fix it?

rjmccall commented:
You almost certainly don't need to modify anything in LLDB; instead, there's probably something broken in the RemoteAST code, i.e. MetadataReader.  I would look at the test case, try to figure out what function type metadata is being parsed, and then try to duplicate that failure in the remoteast test harness.

xedin commented:
@rjmccall Could be it as simple as lldb requiring a clean build instead of incremental to resolve this?

xedin commented:
I've tried using `swift-remoteast-test` with `printType` as in one of the remoteAST tests to print "f" variable and it does print it correctly as `(Int) -> Int` from metadata...
